First update of KEB docs (#35)
kyma-project/kyma-environment-broker: docs/03-02-runtime-components.md

# SAP BTP, Kyma runtime components

> > **NOTE:** Once all Kyma components become independent modules, KEB will no longer be required to send components to Reconciler and this document will be deprecated as irrelevant.

Reviewer: ```suggestion
> **NOTE:** Once all Kyma components become independent modules, Kyma Environment Broker (KEB) will no longer be required to send components to Reconciler and this document will be deprecated as irrelevant.
```

Kyma Environment Broker (KEB) serves the functionality of composing the list of components that are installed in a SAP BTP, Kyma runtime. The diagram and steps describe the KEB workflow in terms of calculating and processing Kyma runtime components:

![runtime-components-architecture](./assets/runtime-components.svg)

1. During KEB initialization, the broker reads two files that contain lists of components to be installed in a Kyma runtime:  

   * `kyma-installer-cluster.yaml` file with the given Kyma version
   * `managed-offering-components.yaml` file with additional managed components that are added at the end of the base components list

2. The user provisions a Kyma runtime and selects optional components that they want to install.

3. KEB composes the final list of components by removing components that were not selected by the user. It also adds the proper global and components overrides and sends the whole provisioning information to the Runtime Provisioner.

> There is a defined [list of the components and their names](https://github.com/kyma-project/control-plane/blob/main/components/kyma-environment-broker/internal/runtime/components). Use these names in your implementation.

Reviewer: ```suggestion
There is a defined [list of the components and their names](https://github.com/kyma-project/kyma-environment-broker/blob/main/internal/runtime/components/components.go). Use these names in your implementation.
```

Author: Thanks for spotting this one! :)

> ## Disabled components

Reviewer: The link to the disable components, in line 22, is also broken. It should probably lead to https://github.com/kyma-project/kyma-environment-broker/blob/main/internal/runtime/disabled_components.go

Author: Yes, that's right 🙇‍♀️

> * If disabling a given component requires more complex logic, create a new file called `internal/runtime/{compoent-name}_disabler.go` and implement a service which fulfills the following interface:

Reviewer: ```suggestion
* If disabling a given component requires more complex logic, create a new file called `internal/runtime/{component-name}_disabler.go` and implement a service that fulfills the following interface:
```

```go
// OptionalComponentDisabler disables component from the given list and returns a modified list
type OptionalComponentDisabler interface {
	Disable(components internal.ComponentConfigurationInputList) internal.ComponentConfigurationInputList
```

> >**TIP**: Check the [CustomDisablerExample](https://github.com/kyma-project/control-plane/blob/main/components/kyma-environment-broker/internal/runtime/custom_disabler_example.go) as an example of custom service for disabling components.

Reviewer: ```suggestion
>**TIP**: Check the [CustomDisablerExample](https://github.com/kyma-project/kyma-environment-broker/blob/main/internal/runtime/custom_disabler_example.go) as an example of custom service for disabling components.
```

}
```

### Remove an optional component from the disabled components list

> If you want to remove the option to disable components and make them required during SAP BTP, Kyma runtime installation, remove a given entry from the **optionalComponentsDisablers** list in the [`cmd/broker/main.go`](https://github.com/kyma-project/control-plane/blob/main/components/kyma-environment-broker/cmd/broker/main.go) file.

Reviewer: ```suggestion
If you want to remove the option to disable components and make them required during SAP BTP, Kyma runtime installation, remove a given entry from the **optionalComponentsDisablers** list in the [`cmd/broker/main.go`](https://github.com/kyma-project/kyma-environment-broker/blob/main/cmd/broker/main.go) file.
```
